### PR TITLE
ClusterManager; added minClusterSize as option

### PR DESCRIPTION
**Pull request by meierjan**

### CODE REVIEW

**Comment by meierjan** [comment]
Problem:
Many times the cluster only contain 2 or 3 items even if there would - for some tastes - be enough space to display them individually.

Solution:
The `ClusterManager` can now receive a 'minClusterSize' so that it will not start clustering until a certain amount of items is reached.

**Comment by meierjan** [comment]
Related to issue #7

**Comment by makovkastar** [comment]
Thank you, I will take a look as soon as I have time.

File: library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java [hunk 1]
```diff
@@ -109,6 +112,16 @@ public void setItems(@NonNull List<T> clusterItems) {
...
     }
 
+    /**
+     * Sets the minimum cluster-size. Clusters with less will be broken-up into single-item clusters
+     */
+    public void setMinClusterSize(int minClusterSize) {
+        if (minClusterSize < 0) {
+            throw new IllegalArgumentException("minClusterSize must be positive");
+        }
+        this.mMinClusterSize = minClusterSize;
```
Comment (makovkastar) [inline library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java]: No need for `this`.

File: library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java [hunk 2]
```diff
@@ -109,6 +112,16 @@ public void setItems(@NonNull List<T> clusterItems) {
         buildQuadTree(clusterItems);
     }
 
+    /**
+     * Sets the minimum cluster-size. Clusters with less will be broken-up into single-item clusters
+     */
+    public void setMinClusterSize(int minClusterSize) {
+        if (minClusterSize < 0) {
+            throw new IllegalArgumentException("minClusterSize must be positive");
+        }
```
Comment (makovkastar) [inline library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java]: I would add a new method to `Preconditions.java`:

```
 static void checkArgument(boolean expression) {
    if (!expression) {
      throw new IllegalArgumentException();
    }
  }
```

And then use it:

```
checkArgument(minClusterSize > 0);
mMinClusterSize = minClusterSize;
```

File: library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java [hunk 3]
```diff
@@ -133,7 +146,7 @@ private void cluster() {
 
     @NonNull
     private List<Cluster<T>> getClusters(@NonNull LatLngBounds latLngBounds, float zoomLevel) {
-        List<Cluster<T>> clusters = new ArrayList<>();
+        final List<Cluster<T>> clusters = new ArrayList<>();
```
Comment (makovkastar) [inline library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java]: We don't use final for local variables in this library.

File: library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java [hunk 4]
```diff
@@ -38,6 +39,8 @@
 
     private AsyncTask mClusterTask;
 
+    private int mMinClusterSize = 1;
```
Comment (makovkastar) [inline library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java]: I prefer to have a constant `DEFAULT_MIN_CLUSTER_SIZE = 1` and then `private int mMinClusterSize = DEFAULT_MIN_CLUSTER_SIZE`.

File: library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java [hunk 5]
```diff
@@ -162,19 +175,44 @@ private void cluster() {
...
+                        clusters.add(
+                                new Cluster<>(
+                                        latitude,
+                                        longitude,
+                                        points,
+                                        north,
+                                        west,
+                                        south,
+                                        east
+                                )
```
Comment (makovkastar) [inline library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java]: I prefer the formatting that was here before.

File: library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java [hunk 6]
```diff
@@ -162,19 +175,44 @@ private void cluster() {
...
+                                    new Cluster<>(
+                                            point.getLatitude(),
+                                            point.getLongitude(),
+                                            Collections.singletonList(point),
+                                            north,
+                                            west,
+                                            south,
+                                            east
+                                    )
+                            );
```
Comment (makovkastar) [inline library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java]: I prefer the formatting that was here before.

File: library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java [hunk 7]
```diff
@@ -162,19 +175,44 @@ private void cluster() {
...
-                    double totalLongitude = 0;
-
-                    for (QuadTreePoint point : points) {
-                        totalLatitude += point.getLatitude();
-                        totalLongitude += point.getLongitude();
+                    if (points.size() >= mMinClusterSize) {
+                        double totalLatitude = 0;
+                        double totalLongitude = 0;
+
+                        for (QuadTreePoint point : points) {
```
Comment (makovkastar) [inline library/src/main/java/net/sharewire/googlemapsclustering/ClusterManager.java]: Replace QuadTreePoint with T.

**Review by makovkastar — CHANGES_REQUESTED** [review]

**Comment by meierjan** [comment]
There u go.

**Comment by makovkastar** [comment]
Good work! Could you squash the changes into a single commit with a descriptive name and I will merge it?

**Review by makovkastar — APPROVED** [review]

**Comment by meierjan** [comment]
There u go.

**Comment by PJHaynes304** [comment]
This is a great feature! Is there any of chance of a new version being pushed to bintray so that we can use the new feature when using the library through gradle dependency?

**Comment by makovkastar** [comment]
I will push the new version to Bintray as soon as possible.

**Comment by PJHaynes304** [comment]
Appreciate it, thanks!

**Comment by makovkastar** [comment]
The new version 0.1.3 has been released to Bintray.

**Comment by PJHaynes304** [comment]
Great, thanks! just built with the new version and having a play around with the minimum cluster size. @meierjan would you expect this to work when dynamically changed? I'm setting a minimum cluster size at a certain zoom level to essentially disable clustering for items that have duplicate locations. When zoomed back out, after I reassign the default minimum cluster size to 1 the items aren't reclustered. Is this expected behaviour?

**Comment by meierjan** [comment]
Hey @PJHaynes304, I did this intensional. But yeah, I had the same question in mind and I think this should be discusses. I'd like to hear @makovkastar on that as well.

As a workaround I think you can call `ClusterManager#onCameraIdle` after setting the size - if I recall correctly.

**Pull request merged**